WalletWasabi/Wallets/CpfpInfoProvider.cs

```diff
@@ -0,0 +1,218 @@
+		{
+			_httpClient = httpClientFactory.NewHttpClient(() => new Uri("https://mempool.space/api/"), Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+		}
+		else if(network == Network.TestNet)
+		{
+			_httpClient = httpClientFactory.NewHttpClient(() => new Uri("https://mempool.space/testnet/api/"), Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+		}
+		else
+		{
+			throw new InvalidOperationException("CpfpInfoProvider is only operational on Main or TestNet");
+		}
+	}
+
+
+	private readonly IHttpClient _httpClient;
+
+
+	private readonly Channel<SmartTransaction> _transactionsChannel = Channel.CreateUnbounded<SmartTransaction>();
+	private readonly Dictionary<uint256, CachedCpfpInfo> _cpfpInfoCache = new();
+	private AsyncLock AsyncLock { get; } = new();
+
+	private DateTimeOffset _lastUpdateCacheLoop = DateTimeOffset.MinValue;
+
+	public event EventHandler<EventArgs>? RequestedCpfpInfoArrived;
+
+
+	protected override async Task ExecuteAsync(CancellationToken cancel)
+	{
+		List<uint256> scheduledRequests = [];
+		List<Task> tasks = [];
+		while (!cancel.IsCancellationRequested)
+		{
+			var txToFetch = await _transactionsChannel.Reader.ReadAsync(cancel).ConfigureAwait(false);
+			var txid = txToFetch.GetHash();
+
+			using (await AsyncLock.LockAsync(cancel).ConfigureAwait(false))
+			{
+				if (scheduledRequests.Contains(txid) ||
+				    !ShouldRequest(txToFetch) ||
+				    (_cpfpInfoCache.TryGetValue(txid, out var cachedCpfpInfo) &&
+				     cachedCpfpInfo.TimeLastUpdate > DateTimeOffset.UtcNow - MinimumBetweenUpdateRequests))
+				{
+					// We already scheduled fetch or fetching doesn't make sense anymore or last result is recent enough.
+					continue;
+				}
+			}
+
+			scheduledRequests.Add(txid);
+			tasks.Add(Task.Run(() => ScheduleTask(txToFetch), cancel));
+
+			if (tasks.Count <= MaximumScheduledRequestsInParallel)
+			{
+				continue;
+			}
+
+			Task completedTask = await Task.WhenAny(tasks).ConfigureAwait(false);
+			tasks.Remove(completedTask);
+		}
+
+		async Task ScheduleTask(SmartTransaction transaction)
+		{
+			var random = new Random();
+			var delayInSeconds = random.Next(MaximumDelayInSeconds);
+			var delay = TimeSpan.FromSeconds(delayInSeconds);
+
+			try
+			{
+				await Task.Delay(delay, cancel).ConfigureAwait(false);
+			}
+			catch (OperationCanceledException)
+			{
+				Logger.LogTrace($"FetchCpfpInfoAsync was unscheduled for {transaction.GetHash()} because Wasabi is shutting down");
+				return;
+			}
+
+			try
+			{
+				await FetchCpfpInfoAsync(transaction, cancel).ConfigureAwait(false);
+				scheduledRequests.Remove(transaction.GetHash());
+			}
+			catch (Exception ex) when (ex is HttpRequestException or OperationCanceledException)
+			{
+				if (cancel.IsCancellationRequested)
+				{
+					Logger.LogTrace($"FetchCpfpInfoAsync was canceled for {transaction.GetHash()} because Wasabi is shutting down");
+					return;
+				}
+
+				// Reschedule
+				if (ShouldRequest(transaction))
```

Comment:
As stated above we are not rescheduling in case of an exception other than `HttpRequestException` and `OperationCanceledException`. In fact it's an unobserved exception because nothing will throw (as we are not awaiting the result), but the request will be cancelled.

The request not being removed was an error, I added it to the finally because even if it makes the code slightly unoptimized (it tries to remove several times in case of a rescheduling) it is easier to understand that way.